plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs

```diff
@@ -0,0 +1,225 @@
+                | otherwise -> maybe ty id $ Map.lookup x sub
+      TyFun _ a b      -> TyFun () (go fvs seen sub a) (go fvs seen sub b)
+      TyApp _ a b      -> TyApp () (go fvs seen sub a) (go fvs seen sub b)
+      TyLam _ x k b
+        | Set.member x fvs -> TyLam () x' k $ go (Set.insert x' fvs) seen sub (renameType x x' b)
+        | otherwise        -> TyLam () x  k $ go (Set.insert x fvs) (Set.delete x seen) sub b
+        where x' = freshenTyName (fvs <> setOf ftvTy b) x
+      TyForall _ x k b
+        | Set.member x fvs -> TyForall () x' k $ go (Set.insert x' fvs) seen sub (renameType x x' b)
+        | otherwise        -> TyForall () x  k $ go (Set.insert x fvs) (Set.delete x seen) sub b
+        where x' = freshenTyName (fvs <> setOf ftvTy b) x
+      TyBuiltin{}      -> ty
+      TyIFix _ a b     -> TyIFix () (go fvs seen sub a) (go fvs seen sub b)
+
+-- | Find all free type variables of type `a` given substitution `sub`. If variable `x` is
+-- free in `a` but in the domain of `sub` we look up `x` in `sub` and get all the free type
+-- variables of the result - up to the substitution.
+fvTypeR :: Map TyName (Type TyName DefaultUni ()) -> Type TyName DefaultUni () -> Set TyName
+fvTypeR sub a = Set.unions $ freeAndNotInSub : map (fvTypeR sub . (Map.!) sub) (Set.toList freeButInSub)
+      where
+          fvs = setOf ftvTy a
+          subDom = Map.keysSet sub
+          freeButInSub = Set.intersection subDom fvs
+          freeAndNotInSub = Set.difference fvs subDom
+
+-- * Generators for substitutions
+
+-- | Generate a type substitution that is valid in a given context.
+genSubst :: Map TyName (Kind ()) -> Gen (Map TyName (Type TyName DefaultUni ()))
+genSubst ctx = do
+  xks <- sublistOf <=< shuffle $ Map.toList ctx
+  go ctx Map.empty xks
+  where
+    -- Counts is used to balance the ratio between the number
+    -- of times a variable `x` appears in the substitution and
+    -- the size of the type it maps to - the more times `x` appears
+    -- the smaller the type it maps to needs to be to avoid blowup.
+    go _ _ [] = return mempty
+    go ctx counts ((x, k) : xs) = do
+      let ctx' = Map.delete x ctx
+          w    = fromMaybe 1 $ Map.lookup x counts
+      ty <- sized $ \ n -> resize (div n w) $ genTypeWithCtx ctx' k
+      let moreCounts = fmap (* w) $ fvTypeBag ty
+          counts'    = Map.unionWith (+) counts moreCounts
+      Map.insert x ty <$> go ctx' counts' xs
+
+
+shrinkSubst :: Map TyName (Kind ())
+            -> Map TyName (Type TyName DefaultUni ())
+            -> [Map TyName (Type TyName DefaultUni ())]
+shrinkSubst ctx = map Map.fromList . liftShrink shrinkTy . Map.toList
+  where
+    shrinkTy (x, ty) = (,) x <$> shrinkTypeAtKind (pruneCtx ctx ty) k ty
+      where k = fromMaybe (error $ "internal error: " ++ show x ++ " not found") $ Map.lookup x ctx
+    pruneCtx ctx ty = Map.filterWithKey (\ x _ -> Set.member x fvs) ctx
```

Comment:
There's [`restrictKeys`](https://github.com/haskell/containers/blob/50175b72dc781f82a419bddafba1bdd758fbee4b/containers/src/Data/Map/Internal.hs#L1984) that has this semantics, but is faster. I think it can be used both here and in `fvTypeR`. I'll try that out.